ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Forging.hs

```diff
@@ -31,195 +30,182 @@ import           Ouroboros.Consensus.HardFork.Combinator.Ledger (Ticked (..))
+        -- enforce it statically.
+      $ matchNS' "ForgeStateInfo"       (getOneEraForgeStateInfo forgeStateInfo)
+      $ matchNS' "IsLeader"             (getOneEraIsLeader isLeader)
+      $ matchNS' "Ticked ChainDepState" (State.tip chainDepState)
+      $ blockForging
   where
     distrib ::
          NS (Maybe :.: WrapCannotForge) xs
       -> Either (HardForkCannotForge xs) ()
     distrib = maybe (Right ()) (Left . OneEraCannotForge) . hsequence'
 
     checkOne ::
-         (Maybe :.: BlockForging m) blk
-      -> TopLevelConfig blk
-      -> (Maybe :.: WrapForgeStateInfo) blk
-      -> Product WrapIsLeader (Ticked :.: WrapChainDepState) blk
+         TopLevelConfig blk
+      -> Product
+           WrapForgeStateInfo
+           (Product
+             WrapIsLeader
+             (Product
+               (Ticked :.: WrapChainDepState)
+               (BlockForging m)))
+           blk
       -> (Maybe :.: WrapCannotForge) blk
          -- ^ We use @Maybe x@ instead of @Either x ()@ because the former can
          -- be partially applied.
-    checkOne (Comp mBlockForging)
-             cfg'
-             (Comp mForgeStateInfo')
-             (Pair (WrapIsLeader isLeader') (Comp tickedChainDepState)) =
+    checkOne cfg'
+             (Pair
+               (WrapForgeStateInfo forgeStateInfo')
+               (Pair
+                 (WrapIsLeader isLeader')
+                 (Pair
+                   (Comp tickedChainDepState)
+                   blockForging'))) =
         Comp $ either (Just . WrapCannotForge) (const Nothing) $
           checkCanForge
-            blockForging
+            blockForging'
             cfg'
             curSlot
             (unwrapTickedChainDepState tickedChainDepState)
             isLeader'
             forgeStateInfo'
-      where
-        blockForging = case mBlockForging of
-          Just bf -> bf
-          -- We are given an 'IsLeader' proof of the era while we don't have a
-          -- 'BlockForging' record for that era, impossible
-          Nothing ->
-            error "checkCanForge in an era in which we cannot lead"
-
-        forgeStateInfo' = case mForgeStateInfo' of
-          Just (WrapForgeStateInfo fsi) -> fsi
-          -- We are given an 'IsLeader' proof of the era, so we must have a
-          -- 'BlockForging' record for the same era. We are also given
-          -- 'ForgeStateInfo's for all eras for which we have a 'BlockForging'
-          -- record, so it must be that there is a 'ForgeStateInfo' for this
-          -- era.
-          Nothing ->
-            error "checkCanForge in an era in which we cannot lead"
 
 hardForkForgeBlock ::
      forall m xs. (CanHardFork xs, Monad m)
-  => OptNP 'False (BlockForging m) xs
+  => NS (BlockForging m) xs
   -> TopLevelConfig (HardForkBlock xs)
   -> BlockNo
   -> SlotNo
   -> TickedLedgerState (HardForkBlock xs)
   -> [GenTx (HardForkBlock xs)]
   -> HardForkIsLeader xs
   -> m (HardForkBlock xs)
```

Comment:
Precondition: `NS BlockForging`, ticked ledger state, and `HardForkIsLeader` are aligned. This follows from the postcondition of `check(IsLeader)` and the fact that the ticked chain dep state and ticked ledger state are from the same era.